### ISSUE TITLE
Cannot run conda 4.4 in a batch script

### ISSUE THREAD

**Issue by isuruf** [conda/conda#6553]
For example if I run a .bat file with
```
conda --version
echo asd
```
then only the first line is executed.

cc @kalefranz, this is the reason for the error @jdblischak mentioned in https://github.com/conda/conda/issues/6527

**Comment by kalefranz** [comment]
Ok, this is all then related to something we've done in https://github.com/conda/conda/blob/1a3353accd4b782835b9736627f890337ff488c1/conda/shell/Library/bin/conda.bat

**Comment by kalefranz** [comment]
We probably need to add a lot of

    @IF errorlevel 1 exit /b 1

checks to that file.

**Comment by isuruf** [comment]
Windows shell `cmd.exe` works differently than bash. When you run a batch script from one batch script control is given to the child script and the parent is simply stopped. To avoid this and have the child return control to the parent you need to use the command call.

Instead of 
```
conda --version
echo asd
```
you now need to use
```
call conda --version
echo asd
```
otherwise the 2nd line will not be run anymore.

**Comment by kalefranz** [comment]
What if, instead, your bash script was

    conda.exe --version
    echo asd

Is that a possible fix?

**Comment by isuruf** [comment]
Yes, that works, but the rest of conda.bat is not run. Isn't that a problem?

**Comment by isuruf** [comment]
Would conda be supporting using the activate.bat script in the future? If so, calling the activate script and then running conda.exe is okay

**Comment by kalefranz** [comment]
> Would conda be supporting using the activate.bat script in the future?

We want to get rid of the `activate` "executable", because it has name collisions with something like virtualenv's `activate`.  But I agree that we have an unsolved issue here before we can fully end up deprecating it.

> but the rest of conda.bat is not run. Isn't that a problem?

The answer is, it depends.  For now, using `conda.exe` instead of `conda` is no different than what we had in 4.3.  Besides adding real `conda activate` and `conda deactivate` commands, the only other thing `conda.bat` does is call "reactivate" for `install`, `update`, and `remove`.  But conda 4.3 never did "reactivate", so at least you're no worse off using `conda.exe`.

**Comment by mbargull** [comment]
Maybe we could add something like
```bat
@call :run
@goto :eof
:run
```
at the top of `conda.bat` to always run the rest inside a `call`? That just came to mind but I haven't tried it out yet.

**Comment by mbargull** [comment]
I'm not sure about the exit status propagation for this, though...

**Comment by kalefranz** [comment]
I was hoping you might have some ideas here @mbargull :)

**Comment by mbargull** [comment]
nvm, thinking about it, it is exactly as @isuruf said
> When you run a batch script from one batch script control is given to the child script and *the parent is simply stopped*. To avoid this and have the child return control to the parent you need to use the command call.

So I'm not sure we can use a batch file-only solution if we want to make `conda ...` work without `call` in batch files :/ .. Am thinking about it some more time...

**Comment by kalefranz** [comment]
We're probably just stuck with the constraints of batch scripting and cmd.exe.  Eventually when we have the conda.bat pattern more broadly established, maybe we can just switch everybody writing batch scripts to use `call conda`.  In the interim, while people are flipping back and forth between 4.3 and 4.4, `conda.exe` seems like the best solution.

**Comment by mbargull** [comment]
Yeah, I can't come up with a solution which supports the non-`call` usage but uses `conda.bat`..
A question might be whether you would might just want to deprecate the old usage but support it until some `4.5` version to allow a transition phase. That would mean to not add `shell/Library/bin` to path on default for Windows, i.e., withheld the `4.4` shell function/`conda activate` feature on Windows for some time.

**Comment by mbargull** [comment]
Ah, I just didn't look at the install scripts and thus didn't notice
https://github.com/conda/conda/blob/4.4.3/utils/functions.sh#L192-L196
and thought `conda.bat` would still be inside `shell` (with `shell` being added to also added to `$PATH`, too). Removing `Library/bin` from `$PATH` would of course be no option and moving `conda.bat` to `bin` doesn't sound good to me, likewise. Currently, I don't have any further ideas :/ ...
(Stupid `CMD`! Let's just dream of an utopia with world peace and a decommissioned `CMD`... dammit!)

**Comment by mingwandroid** [comment]
Well Powershell is deprecating cmd so I wonder if we shouldn't just do the best we can there (I know nothing about Powershell though) and support only using call for cmd?

**Comment by kalefranz** [comment]
All the other shell wrappers are working out pretty well I think.  Even the `conda.bat` isn't that bad, although this is certainly and issue.  So is, the question of what we do when we deprecate activate.bat (I have ideas there).

Actually, the only wrapper that's still pending to go into 4.4 is the powershell one.  I have NO experience there though, and really don't know the best practices "powershell way" on how to implement that.  At least there are plenty of examples now in the `conda/shell` directory for all the other wrappers.

**Comment by mbargull** [comment]
> and support only using call for cmd? (@mingwandroid)

Of course, I agree you should
> just switch everybody writing batch scripts to use `call conda` (@kalefranz)

as soon as possible since I think the more unified shell handling with the wrappers, including `conda.bat`, is a really nice step forward! If it's possible to create a transitioning period, that would be nice, I guess, but if it's not easily done, pushing the unified concept should be more important, IMHO.
(As for Powershell, absolutely no experience here either -- scurried away from Windows before that became more widely known.)

**Comment by kalefranz** [comment]
> scurried away from Windows before that became more widely known

I just installed powershell on my mac.  😱

`pwsh`

Yes, they changed the executable name for unix from what it is on Windows.  Microsoft maybe just won't ever learn.

**Comment by isuruf** [comment]
Maybe this should go in another issue, but note that you'll run into trouble updating `conda.bat` in the future as the batch script will be changing itself and it'll error out if you change anything above the invocation to `conda.exe`.

**Comment by kalefranz** [comment]
![9a774a6d-c480-4f7a-81e7-ae605510f965](https://user-images.githubusercontent.com/1418419/34324753-f40eb66e-e842-11e7-8263-47bb1653a4c7.jpeg)

**Comment by kalefranz** [comment]
> Maybe this should go in another issue, but note that you'll run into trouble updating `conda.bat` in the future as the batch script will be changing itself and it'll error out if you change anything above the invocation to `conda.exe`.

I actually think we have this problem all the time now.  We've tried to deal with it multiple times in multiple ways, and where we are right now with it is https://github.com/conda/conda/issues/6631.  Trying again to address the issue in https://github.com/conda/conda/pull/6660, but I've sunk a lot of time into that and still haven't been successful yet.

**Comment by kalefranz** [comment]
I think it's safe to close this issue now.  Let me know if anyone disagrees.

**Comment by certik** [comment]
This really should be documented somewhere. It took me hours to figure out I have to put "call" in front of conda, otherwise the rest of the file is not executed.

**Comment by sytelus** [comment]
I want to share my scripts between Windows and Linux and this issue is breaking it. I hope conda.bat can be eliminated and conda.exe itself executes all necessary code so this hack is not needed.

**Comment by sbourdeauducq** [comment]
Well that's conda for you; you may want to try MSYS2 if you want to get things done.
